src/toolchain.cxx

```diff
@@ -0,0 +1,106 @@
+#include "toolchain.h"
+
+#include <sstream>
+
+ToolChainInvocation::ToolChainInvocation(std::string command, char const* const* cli)
+{
+    this->command = command;
+    this->parse_command_args(cli);
+}
+
+void ToolChainInvocation::interpolateSpackEnv(SpackEnvState &spackenv) {
+    // inject Spack includes before the default includes
+    for( auto &include: spackenv.SpackIncludeDirs )
+    {
+        auto incArg = this->composeIncludeArg(include);
+        this->includeArgs.insert(this->includeArgs.begin(), incArg);
+    }
+    for( auto &lib: spackenv.SpackLdLibs )
+    {
+        this->libArgs.push_back(lib);
+    }
+    this->addExtraLibPaths(spackenv.SpackLinkDirs);
+    this->addExtraLibPaths(spackenv.SpackRPathDirs);
+    this->addExtraLibPaths(spackenv.SpackCompilerExtraRPaths);
+    this->addExtraLibPaths(spackenv.SpackCompilerImplicitRPaths);
+    this->loadToolchainDependentSpackVars(spackenv);
+}
+
+int ToolChainInvocation::invokeToolchain() {
+    StrList commandLine(this->composeCommandLists({
+        this->commandArgs,
+        this->includeArgs,
+        this->libArgs,
+        this->libDirArgs,
+        this->objArgs
+    }));
+
+    this->executor = ExecuteCommand(  this->spackCommand,
+                                      commandLine
+                                    );
+    try{
+        // Run first pass of command as requested by caller
+        this->executor.execute();
+    }
+    catch(SpackException &e){
+        std::cerr << e.what() << "\n";
+        return 0;
+    }
+    return 1;
+}
+
+void ToolChainInvocation::parse_command_args(char const* const* cli) {
+    // Collect include args as we need to ensure Spack
+    // Includes come first
+    for( char const* const* c = cli; *c; c++ ){
+        std::string arg = std::string(*c);
+        if ( startswith(arg, "/I") ) {
+            // We have an include arg
+            // can have an optional space
+            // check if there are characters after
+            // "/I" and if not we consider the next
+            // argument to be the include
+            if (arg.size() > 2)
+                this->includeArgs.push_back(arg);
+            else
+                this->includeArgs.push_back(std::string(*(++c)));
+        }
+        else if( endswith(arg, ".lib") )
+            // Lib args are just libraries
+            // provided like system32.lib on the
+            // command line.
+            // lib specification order does not matter
+            // on MSVC but this is useful for filtering system libs
+            // and adding all libs
+            this->libArgs.push_back(arg);
+        else if ( endswith(arg, ".obj") )
+            this->objArgs.push_back(arg);
+        else
+            this->commandArgs.push_back(arg);
+    }
+}
+
+std::string ToolChainInvocation::composeIncludeArg(std::string &include) {
+    return "/I" + include;
+}
```

Comment:
(minor) for `lib/env/cc`, we turn all spack includes into `-isystem` because of the desired precedence semantics when combined with `-I`. It looks like there is no analog for `cl`, and it might be worth noting that here.

Reply:
Ahh right! Thanks for reminding me, definitely worth documenting!